### ISSUE TITLE
unstable tests

### ISSUE THREAD

**Issue by pjungermann** [backstage/backstage#13965]
## Expected Behavior

Tests succeed

## Actual Behavior

Tests fail on unrelated changes (sometimes).

```
Summary of all failing tests
FAIL plugins/catalog-backend/src/service/DefaultRefreshService.test.ts (73.403 s)
  ● Refresh integration › should refresh the parent location, "POSTGRES_13"

    Failed to connect to the database to make sure that 'backstage_plugin_dbe08be3e065a28f29a4be0514db7284d1' exists, KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?

      264 |    * client type.
      265 |    *
    > 266 |    * @param pluginId - The plugin that the database config should correspond with
          |               ^
      267 |    */
      268 |   private getConfigForPlugin(pluginId: string): Knex.Config {
      269 |     const { client } = this.getClientType(pluginId);

      at DatabaseManager.getDatabase (../../../packages/backend-common/src/database/DatabaseManager.ts:266:15)
      at TestDatabases.init (../../../packages/backend-test-utils/src/database/TestDatabases.ts:142:24)
      at createDatabase (service/DefaultRefreshService.test.ts:46:18)
      at service/DefaultRefreshService.test.ts:179:28

  ● Refresh integration › should refresh the parent location, "POSTGRES_9"

    Failed to connect to the database to make sure that 'backstage_plugin_dbc39d3c6c51d333b90e067b59abf1076f' exists, KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?

      264 |    * client type.
      265 |    *
    > 266 |    * @param pluginId - The plugin that the database config should correspond with
          |               ^
      267 |    */
      268 |   private getConfigForPlugin(pluginId: string): Knex.Config {
      269 |     const { client } = this.getClientType(pluginId);

      at DatabaseManager.getDatabase (../../../packages/backend-common/src/database/DatabaseManager.ts:266:15)
      at TestDatabases.init (../../../packages/backend-test-utils/src/database/TestDatabases.ts:142:24)
      at createDatabase (service/DefaultRefreshService.test.ts:46:18)
      at service/DefaultRefreshService.test.ts:179:28

FAIL plugins/catalog-backend/src/modules/core/DefaultLocationStore.test.ts (40.517 s)
  ● DefaultLocationStore › should do a full sync with the locations on connect, "POSTGRES_13"

    Failed to connect to the database to make sure that 'backstage_plugin_db6c06ac7e871f9f6de97ac05177610540' exists, KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?

      264 |    * client type.
      265 |    *
    > 266 |    * @param pluginId - The plugin that the database config should correspond with
          |               ^
      267 |    */
      268 |   private getConfigForPlugin(pluginId: string): Knex.Config {
      269 |     const { client } = this.getClientType(pluginId);

      at DatabaseManager.getDatabase (../../../packages/backend-common/src/database/DatabaseManager.ts:266:15)
      at TestDatabases.init (../../../packages/backend-test-utils/src/database/TestDatabases.ts:142:24)
      at createLocationStore (modules/core/DefaultLocationStore.test.ts:27:18)
      at modules/core/DefaultLocationStore.test.ts:38:30
```

Happened to me on two different PRs. It were always the same. Eventually, they succeeded without any code change.

## Steps to Reproduce

1. provide a PR
2. test run fails on CI / test (14.x)

example: https://github.com/backstage/backstage/actions/runs/3164107907/jobs/5152198397

## Your Environment

- Browser Information: 

- Output of `yarn backstage-cli info`: 

```text

```

**Comment by Rugvip** [comment]
indeed! 👍

It started happening since #13574, where test are now running with a single Jest execution rather than through Lerna. The parallelization should be similar, but the test isolation may be affected as test will share the same Jest worker more often.

Been digging into it a bit and tried some workarounds, but haven't found anything that seems to help yet. I suspect that it's not actually related to connection pool limits since #13798 didn't seem to help much. My hunch at this point is that we have some tests with some very bad leaks, of the kind that for example keep catalog processing happening in the background.

**Comment by Parsifal-M** [comment]
> indeed! 👍
> 
> It started happening since #13574, where test are now running with a single Jest execution rather than through Lerna. The parallelization should be similar, but the test isolation may be affected as test will share the same Jest worker more often.
> 
> Been digging into it a bit and tried some workarounds, but haven't found anything that seems to help yet. I suspect that it's not actually related to connection pool limits since #13798 didn't seem to help much. My hunch at this point is that we have some tests with some very bad leaks, of the kind that for example keep catalog processing happening in the background.

Hey, interesting read, how would you find which tests have leaks? Would they be the ones failing? I guess not necessarily... 🤔

**Comment by freben** [comment]
We are honestly open to suggestions on strategies to find and not reintroduce leaks like this.

**Comment by Parsifal-M** [comment]
> If you scroll up a bit in the log linked above, there's a line saying
> 
> > A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --detectOpenHandles to find leaks. Active timers can also cause this, ensure that .unref() was called on them.
> 
> Maybe that flag could give hints.

Ah sorry, I missed that bit :sweat_smile: 

I will do a bit of digging on this, I have some spare time this week. :+1:

**Comment by Rugvip** [comment]
Some fixes in #14170

Found by running `DEBUG=knex:* yarn test ...` towards different tests and finding repeated patterns in the knex logs. Then verifying leakage in individual tests by throwing this at the end and checking whether there is still DB activity:

```ts
it('waits', async () => {
  console.log('##### WAITING #####');
  return new Promise(resolve => setTimeout(resolve, 20000));
}, 30000);
```
